vault/identity_store_oidc_provider.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+type assignment struct {
+	name     string
+	Groups   []string `json:"groups"`
+	Entities []string `json:"entities"`
+}
+
+const (
+	oidcProviderPrefix = "oidc_provider/"
+	assignmentPath     = oidcProviderPrefix + "named_assignments/"
+)
+
+func oidcProviderPaths(i *IdentityStore) []*framework.Path {
+	return []*framework.Path{
+		{
+			Pattern: "oidc/assignment/" + framework.GenericNameRegex("name"),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the assignment",
+				},
+				"entities": {
+					Type:        framework.TypeCommaStringSlice,
+					Description: "Comma separated string or array of identity entity names",
+				},
+				"groups": {
+					Type:        framework.TypeCommaStringSlice,
+					Description: "Comma separated string or array of identity group names",
+				},
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.UpdateOperation: &framework.PathOperation{
+					Callback: i.pathOIDCCreateUpdateAssignment,
+				},
+				logical.CreateOperation: &framework.PathOperation{
+					Callback: i.pathOIDCCreateUpdateAssignment,
+				},
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: i.pathOIDCReadAssignment,
+				},
+				logical.DeleteOperation: &framework.PathOperation{
+					Callback: i.pathOIDCDeleteAssignment,
+				},
+			},
+			ExistenceCheck:  i.pathOIDCAssignmentExistenceCheck,
+			HelpSynopsis:    "CRUD operations for OIDC assignments.",
+			HelpDescription: "Create, Read, Update, and Delete OIDC assignments.",
+		},
+		{
+			Pattern: "oidc/assignment/?$",
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ListOperation: &framework.PathOperation{
+					Callback: i.pathOIDCListAssignment,
+				},
+			},
+			HelpSynopsis:    "List OIDC assignments",
+			HelpDescription: "List all configured OIDC assignments in the identity backend.",
+		},
+	}
+}
+
+// pathOIDCCreateUpdateAssignment is used to create a new assignment or update an existing one
+func (i *IdentityStore) pathOIDCCreateUpdateAssignment(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	var assignment assignment
+	if req.Operation == logical.UpdateOperation {
+		entry, err := req.Storage.Get(ctx, assignmentPath+name)
+		if err != nil {
+			return nil, err
+		}
+		if entry != nil {
+			if err := entry.DecodeJSON(&assignment); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if entitiesRaw, ok := d.GetOk("entities"); ok {
+		assignment.Entities = entitiesRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		assignment.Entities = d.Get("entities").([]string)
+	}
+
+	if groupsRaw, ok := d.GetOk("groups"); ok {
+		assignment.Groups = groupsRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		assignment.Groups = d.Get("groups").([]string)
+	}
```

Comment:
Can you provide some context on this code? I'm not understanding why the else if is needed.

Reply:
I took a closer look at this and the else if is not needed after all. The only difference I can see is that the "else if" block created an empty slice. When then "else if" block is omitted `assignment.Groups` results in a nil slice so I had to update the `deep.Equal` in the UT.

Reply:
Hmm, this is pretty common pattern for create/update handlers. Having a look to see if there is a good reason for it.

Reply:
I took a look around to confirm that this pattern is quite common in create/update type handlers. The `else if` branch handles the case where it's a create operation and the field was not provided. Its purpose is to set the [default or zero value](https://github.com/hashicorp/vault/blob/main/sdk/framework/field_data.go#L66-L71) for the field via the `d.Get()` call. I'm thinking it's okay to leave this as it is instead of removing the `else if` branch and returning a nil slice. Setting the proper default or zero value would be most consistent with the rest of our APIs. 

That said, I agree that reading this code is a bit confusing. I'm open to a simplification if we can find one. I think this would make the purpose of the `else if` branch a bit more clear without changing the behavior:
```
if groupsRaw, ok := d.GetOk("groups"); ok {
        assignment.Groups = groupsRaw.([]string)
} else if req.Operation == logical.CreateOperation {
        assignment.Groups = d.GetDefaultOrZero("groups").([]string)
}
```